Update the tutorial to Tokio 0.2 (#352)
tokio-rs/website: content/docs/getting-started/hello-world.md

the stream and then yields it once it's been created for additional processing.

```rust
use tokio::net::TcpStream;

fn main() {
    let addr = "127.0.0.1:6142";
    let client = TcpStream::connect(addr);

    let future = async move {
        match client.await {
            Ok(_stream) => {
                println!("created stream");

                // Process stream here.
            }
            Err(err) => {

>                 // In our example, we are only going to log the error to STDOUT.

Reviewer: Maybe this should have a comment saying that the `TcpStream::connect` future has an `Output` type of `io::Result<TcpStream>`? Previously, there was a comment about `Future::Error`, but since those no longer exist, we should explain where the error is coming from here.

                println!("connection error = {:?}", err);
            }

>         };

Reviewer: We might want to consider modifying this code to just use the `?` operator — removing error handling boilerplate would let users focus on the new content being introduced. Maybe something like:
```rust
let future = async move {
    let stream = client.await?;
    println!("created stream");
# drop(stream); // this lets the `stream` local appear to be used so we don't need `_stream`

    // process stream here

    Ok(())
};
```
and add a comment explaining that we can use `?` in an `async` block the way we can in other code.

    };

> }

Reviewer: In this example, the future is never run. I think that, in the previous example, that is okay, since we hid the `fn main()` and it just looked like code somewhere inside a program. Here, though, we are showing `fn main()` at the top of the example. IMO, this gives the incorrect implication that users can write code like this and expect it to work.

Maybe we should hide the `fn main()` in this example?

```

> Inside the block the call to `TcpStream::connect` returns a [`Future`] of the created TCP stream.

Reviewer: The call to `TcipStream::connect` is not inside an `async` block here?

```suggestion
The call to `TcpStream::connect` returns a [`Future`] of the created TCP stream.
```

We'll learn more about [`Futures`] later in the guide, but for now you can think of
a [`Future`] as a value that represents something that will eventually happen in the
future (in this case the stream will be created). This means that `TcpStream::connect` does

> not wait for the stream to be created before it returns. Rather it returns immediately

Reviewer: Also, I know this was already here, but I think there should be a comma after "Rather"
```suggestion
not wait for the stream to be created before it returns. Rather, it returns immediately
```

with a value representing the work of creating a TCP stream.

> Note the `async`-block that follows. Contents of the async block are not executed immediately but are rather

Reviewer: Some minor nits:
```suggestion
Note the `async` block that follows. Contents of the `async` block are not executed immediately, but are rather
```

Reviewer: Also, we use "rather" in the last sentence of the previous paragraph. What about something like:
```suggestion
Note the `async` block that follows. Like the `TcpStream::connect` future, the contents of the `async` block are not executed immediately. Instead, they are
```

> stored in an object whose result can be computed later. Inside the `async` block we can `await` on the

Reviewer: ```suggestion
stored in an object whose result can be computed later. Inside the `async` block, we can `await` the
```

> created future so that any code that follows is executed with the result of the `await`-ed [`Future`]'s computation.

Reviewer: ```suggestion
created future, so that any code that follows is executed with the result of the `await`-ed [`Future`]'s computation.
```

We will be digging more into futures (and the related concepts of streams and sinks)
later on.

Next, we will process the stream.

# Writing data

Our goal is to write `"hello world\n"` to the stream.

> Going back to the `TcpStream::connect(addr)` block:

Reviewer: `TcpStream::connect(addr)` isn't in a block, how about:
```suggestion
Going back to the `async` block that `await`s the `TcpStream::connect` future:
```
or just
```suggestion
Going back to the `async` block in the previous example:
```

```rust
use tokio::io::AsyncWriteExt;
use tokio::net::TcpStream;

fn main() {
    let addr = "127.0.0.1:6142";
    let client = TcpStream::connect(addr);

    let future = async move {
        if let Ok(mut stream) = client.await {
            println!("created stream");

            let result = stream.write_all(b"hello world\n").await;

            println!("wrote to stream; success={:?}", result.is_ok());

>         }

Reviewer: Before we added the call to `write_all`, this code used a `match`, but now it uses an `if let`...why not continue using the match? That way, we still show error handling, and less code changes between the two examples.

[Tokio runtime][rt].

```rust
use tokio::io::AsyncWriteExt;
use tokio::net::TcpStream;
use tokio::runtime::Runtime;

fn main() {
    let addr = "127.0.0.1:6142";
    let client = TcpStream::connect(addr);

    let future = async move {
        if let Ok(mut stream) = client.await {
            println!("created stream");

            let result = stream.write_all(b"hello world\n").await;

            println!("wrote to stream; success={:?}", result.is_ok());
        }
    };

    let rt = Runtime::new().unwrap();
    println!("About to create the stream and write to it...");
    rt.block_on(future);

>     println!("Stream has been created and written to.");

Reviewer: Since these are "getting started" docs, maybe we should just use the batteries-included `#[tokio::main]` attribute, rather than creating a `Runtime` manually (more of an advanced API). I notice that the previous version used `tokio::run`, which, IMO, is more or less the tokio 0.1 equivalent of `#[tokio::main]`, so perhaps we should continue using the simpler API?

}
```

In this case we create the default runtime, and spawn the task onto it, blocking the current thread until all spawned tasks
have completed and all resources (like files and sockets) have been dropped.

So far, we only have a single task running on the executor, so the `client` task
is the only one blocking `run` from returning. Once `run` has returned we can be sure
that our Future has been run to completion.

This example can also be greatly simplified using the `#[tokio::main]` procedural macro.

```rust
use tokio::io::AsyncWriteExt;
use tokio::net::TcpStream;

#[tokio::main]
async fn main() {
    let addr = "127.0.0.1:6142";
    let client = TcpStream::connect(addr);

    if let Ok(mut stream) = client.await {
        println!("created stream");

        let result = stream.write_all(b"hello world\n").await;

        println!("wrote to stream; success={:?}", result.is_ok());
    }
}

> ```

Reviewer: Ah, I see we introduce `#[tokio::main]` here...I still think it _might_ be better to show the easier API first, and then say "okay, here is what `tokio::main` actually _does_" and show code that manually creates a runtime.
